### PR TITLE
fix(compiler): incorrectly matching directives on attribute bindings

### PR DESCRIPTION
**Pull request by crisbeto**
Fixes that the compiler was matching directives based on `attr` bindings which doesn't correspond to the runtime behavior. This wasn't a problem until now because the matched directives would basically be a noop, but they can cause issues with required inputs.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/typecheck/test/diagnostics_spec.ts [hunk 1]
```diff
@@ -1093,6 +1093,31 @@ class TestComponent {
         `TestComponent.html(1, 1): Required input 'customAlias' from directive HostDir must be specified.`
       ]);
     });
+
+    it('should not report missing required inputs for an attribute binding with the same name',
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/typecheck/test/diagnostics_spec.ts:1097]: Drive-by comment: a more direct test in https://github.com/angular/angular/blob/54ceed53e21a6a899617b626becc51ffb1059178/packages/compiler/test/render3/view/binding_spec.ts would be nice.

**Review by atscott — APPROVED** [review]
LGTM, assuming Joost's comment is addressed.

**Comment by AndrewKushnir** [comment]
This PR was merged into the repository by commit 8020347f266116feedeb3ea584ba3f12e921f7b3.
